datacube_wps/processes/__init__.py

```diff
@@ -301,38 +303,35 @@ def query_handler(self, time, feature):
 
         return {'data': df, 'chart': chart}
 
+    @log_call
     def input_data(self, dc, time, feature):
         if time is None:
             bag = self.input.query(dc, geopolygon=feature)
         else:
             bag = self.input.query(dc, time=time, geopolygon=feature)
-        measurements = self.input.output_measurements(bag.product_definitions)
-
-        # TODO group
-        # TODO customize the number of processes
-        driller = PD(16)
 
         lonlat = feature.coords[0]
+        NUM_WORKERS = int(os.getenv('DATACUBE_WPS_NUM_WORKERS', '4'))
 
-        try:
-            dss = bag.bag
-        except AttributeError:
-            # datacube 1.7 compatibility
-            dss = bag.pile
+        measurements = self.input.output_measurements(bag.product_definitions)
+        box = self.input.group(bag)
+
+        data = self.input.fetch(box, dask_chunks={'time': 1})
+
+        with Client(n_workers=NUM_WORKERS, processes=False, threads_per_worker=1) as client:
```

Comment:
do you have a reason to use `n_worker=N, threads_per_worker=1` rather than other way around? You are not launching new processes anyway since `processes=False`

Reply:
Launching new cluster on every response seems wasteful and slow. Why not have one ready and idling?

Reply:
I tried one ready and idling approach, but after processing a couple of requests the subsequent ones seemed to trigger garbage collection a lot more frequently than I anticipated. I understand that this is useful, but closing down the cluster seemed to have done a better job cleaning up. If you have ideas on how to instruct the cluster to clean up that would be really appreciated.

I don't have a reason not to have it the other way around! Thanks for the suggestion, am trying it out.

Reply:
`client.restart()` does restart, but it's probably as costly as starting new one, also you might be sharing it with other requests being processed, so probably not a good idea to do that.

You are already replacing dask array with output of `.compute()` so you should not have any dask objects hanging around.